semanticdb/metacp/src/main/scala/scala/meta/internal/metacp/Javacp.scala

```diff
@@ -0,0 +1,402 @@
+package scala.meta.internal.metacp
+
+import java.nio.file.Files
+import java.nio.file.Path
+import scala.collection.JavaConverters._
+import scala.collection.mutable.ArrayBuffer
+import scala.collection.mutable.ListBuffer
+import scala.meta.internal.metacp.asm._
+import scala.meta.internal.metacp.asm.JavaTypeSignature._
+import scala.meta.internal.semanticdb3.SymbolInformation.{Kind => k}
+import scala.meta.internal.{semanticdb3 => s}
+import scala.tools.asm.ClassReader
+import scala.tools.asm.tree.ClassNode
+import scala.tools.asm.tree.FieldNode
+import scala.tools.asm.tree.MethodNode
+import scala.tools.asm.{Opcodes => o}
+
+object Javacp {
+
+  def sdocument(root: Path, file: Path, scopes: TypeVariableScopes): s.TextDocument = {
+    val bytes = Files.readAllBytes(file)
+    val node = parseClassNode(bytes)
+    val symbols =
+      try ssymbols(node, scopes)
+      catch {
+        case e: TypeVariableScopeResolutionError =>
+          // TODO: implement inner anonymous classes
+          // This error can happen when we process an anonymous clase
+          // that references type parameters from an enclosing inner class.
+          // We currently process class files in lexicographical order which means that an order like this here
+          // - Super.class
+          // - Super$1.class (inside Super$Inner)
+          // - Super$Inner.class
+          // causes the type variables defined in Super$Inner.class to not
+          // be entered when we process Super$1.class.
+          val IsNumber = "\\d+".r
+          val hasNumberEntry = node.name
+            .split("\\$")
+            .exists(IsNumber.findFirstIn(_).isDefined)
+          if (!hasNumberEntry) throw e
+          else Nil
+      }
+    val uri = root.relativize(file).toString
+    s.TextDocument(
+      schema = s.Schema.SEMANTICDB3,
+      uri = uri,
+      symbols = symbols
+    )
+  }
+
+  val javaLanguage = Some(s.Language("Java"))
+
+  def fromJavaTypeSignature(sig: JavaTypeSignature)(implicit scopes: TypeVariableScopes): s.Type =
```

Comment:
I suggest we make passing the environment explicit. I don't think that saving a few characters in 4 places is worth complicating mental data flow analysis, especially when a mutable data structure is involved.

Reply:
Done.